### PR TITLE
feat(experiment): set bytesin log as optional …

### PR DESCRIPTION
**Pull request by frimuchkov**
https://github.com/verdaccio/verdaccio/issues/1912

### CODE REVIEW

**Comment by frimuchkov** [comment]
Async middlewares are really important for us
I understand, removing bytesin could affect something, so, I've added env

**Review by juanpicado — COMMENTED** [review]
Thanks for finding a solution on this topic. My opinion here is rather go with the configuration file.

```
experiments:
#  token: false
#  search: false
  bytes_off : true
```

Maybe in verdaccio 5 we can extend each property in the config with a counterpart environment variable for easy Dockerify stuff. But I think this is safer, still not clear how the bytes would be handled in 5.x. I don't want to introduce an env variable that will create heaches in a future major migration.

In that way we can experiment safely and see what happens. I think your experiment would be useful to get more feedback.

⚠️ just FYI in thehory I was not pushing new features to `4.x` since `5.x` is quite different already and I cannot rebase, but if we keep it small, we can go ahead, all my support on this one.

**Comment by frimuchkov** [comment]
> My opinion here is rather go with the configuration file.

Yep, sure.
It looks better, but with env there were fewer changes.

**Comment by frimuchkov** [comment]
BTW, we've solved it by adding `app.use(bodyParser.json({ strict: false, limit: maxBodySize }));` before all our middleware
And now, this PR isn't so important
But, maybe, it's not acceptable for somebody

**Comment by juanpicado** [comment]
> BTW, we've solved it by adding app.use(bodyParser.json({ strict: false, limit: maxBodySize })); before all our middleware

Unfortunately, It did not work last time https://github.com/verdaccio/verdaccio/issues/1841

**Comment by frimuchkov** [comment]
> > BTW, we've solved it by adding app.use(bodyParser.json({ strict: false, limit: maxBodySize })); before all our middleware
> 
> Unfortunately, It did not work last time #1841

Yep, I see
But we proxy audit endpoint in our balancer, that's why it works in our situation

So, what do you think about merging this PR? Could it be useful for somebody (peoples who want to write async middleware)?

**Comment by juanpicado** [comment]
This afternoon:) after my work will double check but looks good.

File: src/api/middleware.ts [hunk 1]
```diff
@@ -174,96 +174,100 @@ export const LOG_STATUS_MESSAGE = "@{status}, user: @{user}(@{remoteIP}), req: '
...
-  }
+    if (_.isNil(_cookie) === false) {
+      req.headers.cookie = _cookie;
+    }
 
-  if (_.isNil(_cookie) === false) {
-    req.headers.cookie = _cookie;
-  }
+    let bytesin = 0;
+    if (_.get(config, 'experiments.bytesin_off') === true) {
```
Comment (juanpicado) [inline src/api/middleware.ts:205]: This logic is completely correct, it disables the whole bytes in unless is explicitly enabled.  Also better use optional chaining, I'm trying to get rid of `lodash`. 

```
    let bytesin = 0;
    if (_.isNil(config?.experiments?.bytesin_off) || config?.experiments?.bytesin_off === false) {
      req.on('data', function(chunk): void {
        bytesin += chunk.length;
      });
    }
```

File: conf/default.yaml [hunk 2]
```diff
@@ -80,6 +80,8 @@ logs:
 #  token: false
 #  # support for the new v1 search endpoint, functional by incomplete read more on ticket 1732
 #  search: false
+#  # disable writing body size to logs, read more on ticket 2013
+#  bytesin_off: false
```
Comment (juanpicado) [inline conf/default.yaml:84]: I think #2012 describes better than #2013

File: src/api/middleware.ts (hunk 1 above)
Comment (frimuchkov) [reply src/api/middleware.ts:205]: Yep, sure
Optional chaining is awesome

File: conf/default.yaml (hunk 2 above)
Comment (frimuchkov) [reply conf/default.yaml:84]: What do you mean?

File: src/api/middleware.ts (hunk 1 above)
Comment (frimuchkov) [reply src/api/middleware.ts:205]: Oh, sorry
My mistake, logic is absolutely incorrect

File: conf/default.yaml (hunk 2 above)
Comment (juanpicado) [reply conf/default.yaml:84]: upss sorry I meant here https://github.com/verdaccio/verdaccio/issues/1912#issuecomment-678674600 ticket 1912

File: src/api/middleware.ts (hunk 1 above)
Comment (juanpicado) [reply src/api/middleware.ts:205]: that works for me even better.

File: conf/default.yaml (hunk 2 above)
Comment (frimuchkov) [reply conf/default.yaml:84]: Fixed
Amend + force push or new commit? What do you prefer?

**Comment by frimuchkov** [comment]
So, everything has been fixed

**Comment by juanpicado** [comment]
> So, everything has been fixed

opps, conflict due the other PR

**Comment by frimuchkov** [comment]
> > So, everything has been fixed
> 
> opps, conflict due the other PR

Resolved :)

**Review by juanpicado — APPROVED** [review]
LGTM great stuff :clap: @frimuchkov

**Pull request merged**